### PR TITLE
Fix Floating Point Constant Expressions in tao_idl

### LINKED ISSUE #1713: tao_idl fails to compile const double variable assignments with math
### Version

ACE+TAO 6.5.15 & 7.0.4

### Host machine and operating system

Red Hat Enterprise Linux 8.4

### Compiler name and version (including patch level)

gcc 8.4.1 20200928 (Red Hat 8.4.1-1)

### The $ACE_ROOT/ace/config.h file

config-linux.h

### The $ACE_ROOT/include/makeinclude/platform_macros.GNU file

platform_linux.GNU

### Contents of $ACE_ROOT/bin/MakeProjectCreator/config/default.features

(doesn't exist.  I used release tarballs downloaded from this github)

### AREA/CLASS/EXAMPLE AFFECTED:

tao_idl

### Synopsis

6.5.15's and 7.0.4's tao_idl do not compile IDL that 6.5.9's tao_idl accepted

### Description

I compiled ACE+TAO 7.0.4, installed it to /tmp/ACE-7.0.4, and set up paths appropriately.  Then I ran commands as below:
```
user@box /tmp/test $ cat minimal.idl
module Minimal
{
    const double a = 6.0;
    const double b = 3.0;
    const double div = a / b;
    const double mul = a * b;
    const double add = a + b;
    const double sub = a - b;
};
user@box /tmp/test $ tao_idl minimal.idl
Error - /tmp/ACE-7.0.4/bin/tao_idl: "minimal.idl", line 5: coercion failure (Can not dump this type) to double
Error - /tmp/ACE-7.0.4/bin/tao_idl: "minimal.idl", line 6: coercion failure (Can not dump this type) to double
Error - /tmp/ACE-7.0.4/bin/tao_idl: "minimal.idl", line 7: coercion failure (Can not dump this type) to double
Error - /tmp/ACE-7.0.4/bin/tao_idl: "minimal.idl", line 8: coercion failure (Can not dump this type) to double
tao_idl: minimal.idl: found 4 errors
```

Commenting out lines 5-8 causes the compilation to succeed.

I compiled ACE 6.5.15 and its tao_idl behaves the same way.

The ACE+TAO 6.5.9 tao_idl accepts this IDL file and generates reasonable code.

### PR DESCRIPTION
**Pull request by iguessthislldo**
Fixes #1713 and adds some regression tests for constant expressions. Also tried to do some more cleanup and minor fixes related to #840.

### CODE REVIEW

File: TAO/TAO_IDL/ast/ast_expression.cpp [hunk 1]
```diff
@@ -1736,7 +1742,8 @@ AST_Expression::eval_bin_op (AST_Expression::EvalKind ek)
       return nullptr;
     }
 
-  ExprType const expr_type = eval_kind_to_expr_type (ek);
+  ExprType const expr_type = ek == EK_const ?
+    pd_v1->ev()->et : eval_kind_to_expr_type (ek);
```
Comment (mitza-oci) [inline TAO/TAO_IDL/ast/ast_expression.cpp:1746]:
```suggestion
    pd_v1->ev ()->et : eval_kind_to_expr_type (ek);
```
at least 1 more of these below

File: TAO/TAO_IDL/ast/ast_expression.cpp [hunk 2]
```diff
@@ -1839,79 +1857,79 @@ AST_Expression::eval_mod_op (AST_Expression::EvalKind ek)
...
 
-      if (this->pd_v2->ev ()->u.ullval == 0)
-        {
-          delete retval;
-          retval = nullptr;
-          return nullptr;
-        }
+  bool success = false;
+  switch (expr_type)
+  {
```
Comment (mitza-oci) [inline TAO/TAO_IDL/ast/ast_expression.cpp:1874]:
```suggestion
    {
```

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: TAO/tests/IDL_Test/expressions.idl [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  const float sub = a - b;
+};
+
+module DoubleValues {
+  const double a = 6.0;
+  const double b = 3.0;
+  const double div = a / b;
+  const double mul = a * b;
+  const double add = a + b;
+  const double sub = a - b;
```
Comment (jwillemsen) [inline TAO/tests/IDL_Test/expressions.idl:44]: Maybe add long double also?

**Review by jwillemsen — APPROVED** [review]

**Review by mitza-oci — APPROVED** [review]

**Comment by jwillemsen** [comment]
One new warning, see https://buildlogs.remedy.nl/win_vs2019_opendds_debug/

```
[Details] 470>X:\BUILD\DOC_ROOT\TAO\TAO_IDL\ast\ast_expression.cpp(1835,12): warning C4804: '%': unsafe use of type 'bool' in operation [X:\BUILD\DOC_ROOT\TAO\TAO_IDL\TAO_IDL_FE.vcxproj]
```

**Pull request merged**